libc/stat.c

```diff
@@ -0,0 +1,871 @@
+	TEST_ASSERT_EQUAL_INT(modification_time.tv_sec, buffer.st_mtim.tv_sec);
+
+	TEST_ASSERT_EQUAL_INT(0, fstat(temp_fd, &buffer));
+	TEST_ASSERT_EQUAL_INT(currentTime, buffer.st_ctim.tv_sec);
+	TEST_ASSERT_EQUAL_INT(access_time.tv_sec, buffer.st_atim.tv_sec);
+	TEST_ASSERT_EQUAL_INT(modification_time.tv_sec, buffer.st_mtim.tv_sec);
+
+	close(temp_fd);
+	remove(tempPath);
+}
+
+
+TEST(stat_nlink_size_blk_tim, st_dev_ino)
+{
+	struct stat buffer;
+
+	/* We can't check the exact values of st_dev, st_ino because:*/
+	/* st_dev describes the device on which this file resides */
+	/* st_ino contains the file's inode number. */
+
+	TEST_ASSERT_EQUAL_INT(0, stat(path, &buffer));
+	TEST_ASSERT_GREATER_THAN_INT(0, buffer.st_dev);
+	TEST_ASSERT_GREATER_THAN_INT(0, buffer.st_ino);
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(path, &buffer));
+	TEST_ASSERT_GREATER_THAN_INT(0, buffer.st_dev);
+	TEST_ASSERT_GREATER_THAN_INT(0, buffer.st_ino);
+
+	TEST_ASSERT_EQUAL_INT(0, fstat(fd, &buffer));
+	TEST_ASSERT_GREATER_THAN_INT(0, buffer.st_dev);
+	TEST_ASSERT_GREATER_THAN_INT(0, buffer.st_ino);
+}
+
+
+////////////////////////////////////
+
+TEST_SETUP(stat_errno)
+{
+	fd = open(path, O_CREAT | O_RDONLY, 0666);
+}
+
+
+TEST_TEAR_DOWN(stat_errno)
+{
+	close(fd);
+	remove(path);
+}
+
+
+TEST(stat_errno, ebadf)
+{
+	/* EBADF only occur for fstat() */
+	struct stat buffer;
+	int fd_invaild = open("never_existed.txt", O_RDONLY);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, fstat(INT_MAX, &buffer));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, fstat(-1, &buffer));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, fstat(fd_invaild, &buffer));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+}
+
+
+TEST(stat_errno, eloop)
+{
+	/* Couldn't reproduce ELOOP for lstat */
+	struct stat buffer;
+	const char *secSymPath = "test_stat_sec_symlink";
+
+	unlink(symPath);
+	unlink(secSymPath);
+
+	TEST_ASSERT_EQUAL_INT(0, symlink(secSymPath, symPath));
+	TEST_ASSERT_EQUAL_INT(0, symlink(symPath, secSymPath));
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, stat(symPath, &buffer));
+	TEST_ASSERT_EQUAL_INT(ELOOP, errno);
+
+	TEST_ASSERT_EQUAL_INT(0, unlink(symPath));
+	TEST_ASSERT_EQUAL_INT(0, unlink(secSymPath));
+}
+
+
+TEST(stat_errno, enametoolong)
+{
+	struct stat buffer;
+	char tooLongPath[PATH_MAX];
+	memset(tooLongPath, 'a', PATH_MAX - 1);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, stat(tooLongPath, &buffer));
+	TEST_ASSERT_EQUAL_INT(ENAMETOOLONG, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, lstat(tooLongPath, &buffer));
+	TEST_ASSERT_EQUAL_INT(ENAMETOOLONG, errno);
+}
+
+
+TEST(stat_errno, enoent)
+{
+	struct stat buffer;
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, stat("\0", &buffer));
+	TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, lstat("\0", &buffer));
+	TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, stat("Lorem", &buffer));
```

Comment:
Let's do it in a safer way.

```suggestion
	TEST_ASSERT_EQUAL_INT(-1, stat("test_stat_nonexistent_file", &buffer));
```